RealmSwift/RealmKeyedCollection.swift

```diff
@@ -0,0 +1,856 @@
+////////////////////////////////////////////////////////////////////////////
+//
+// Copyright 2021 Realm Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+////////////////////////////////////////////////////////////////////////////
+
+import Foundation
+import Realm
+
+/**
+ A homogenous key-value collection of `Object`s which can be retrieved, filtered, sorted, and operated upon.
+*/
+public protocol RealmKeyedCollection: Sequence, ThreadConfined, CustomStringConvertible {
+    /// The type of key associated with this collection
+    associatedtype Key: _MapKey
+    /// The type of value associated with this collection.
+    associatedtype Value: RealmCollectionValue
+
+    // MARK: Properties
+
+    /// The Realm which manages the map, or `nil` if the map is unmanaged.
+    var realm: Realm? { get }
+
+    /// Indicates if the map can no longer be accessed.
+    var isInvalidated: Bool { get }
+
+    /// Returns the number of key-value pairs in this map.
+    var count: Int { get }
+
+     /// A human-readable description of the objects contained in the Map.
+    var description: String { get }
+
+    // MARK: Mutation
+
+    /**
+     Updates the value stored in the dictionary for the given key, or adds a new key-value pair if the key does not exist.
```

Comment:
Each of these should note that they require a write transaction for managed dictionaries. Things which can add values to the dictionary should note that if the value being added is an unmanaged object the object will be added to the realm.